jerry-debugger/src/lib/__tests__/utils.test.ts

```diff
@@ -0,0 +1,191 @@
+  });
+
+  it('returns two bytes for C with big endian', () => {
+    const array = Uint8Array.from([1, 2, 3]);  // 3 ignored
+    expect(decodeMessage({
+      cpointerSize: 2,
+      littleEndian: false,
+    }, 'C', array)).toEqual([(1 << 8) + 2]);
+  });
+
+  it('returns four bytes for C with default config', () => {
+    const array = Uint8Array.from([1, 2, 3, 4, 5]);  // 5 ignored
+    expect(decodeMessage(altConfig, 'C', array)).toEqual(
+      [1 + (2 << 8) + (3 << 16) + (4 << 24)],
+    );
+  });
+
+  it('returns four bytes for C with big endian', () => {
+    const array = Uint8Array.from([1, 2, 3, 4, 5]);  // 5 ignored
+    expect(decodeMessage({
+      cpointerSize: 4,
+      littleEndian: false,
+    }, 'C', array)).toEqual(
+      [(1 << 24) + (2 << 16) + (3 << 8) + 4],
+    );
+  });
+
+  it('handles multiple format characters', () => {
+    const array = Uint8Array.from([1, 2, 3, 4, 5, 6, 7, 8]);  // 8 ignored
+    expect(decodeMessage(defConfig, 'IBC', array)).toEqual([
+      1 + (2 << 8) + (3 << 16) + (4 << 24),
+      5,
+      6 + (7 << 8),
+    ]);
+  });
+
+  it('throws on unexpected pointer size', () => {
+    const array = Uint8Array.from([0, 1, 2, 3, 4, 5, 6, 7]);
+    expect(() => {
+      decodeMessage({
+        cpointerSize: 6,
+        littleEndian: true,
+      }, 'C', array);
+    }).toThrow();
+  });
+});
+
+describe('cesu8ToString', () => {
+  it('returns empty string for undefined input', () => {
+    expect(cesu8ToString(undefined)).toEqual('');
+  });
+
+  it('returns ASCII from ASCII', () => {
+    const sentence = 'The quick brown fox jumped over the lazy dog.';
+    const array = new Uint8Array(sentence.length);
+    for (let i = 0; i < sentence.length; i++) {
+      array[i] = sentence.charCodeAt(i);
+    }
+    expect(cesu8ToString(array)).toEqual(sentence);
+  });
+
+  it('acts like UTF-8 for two-byte encodings', () => {
+    // 0x080 = 00010 000000 = 0x02, 0x00
+    const lowTwoByte = Uint8Array.from([0xc0 + 0x02, 0x80 + 0x00]);
+    // 0x7ff = 11111 111111 = 0x1f, 0x3f
+    const highTwoByte = Uint8Array.from([0xc0 + 0x1f, 0x80 + 0x3f]);
+    expect(cesu8ToString(lowTwoByte)).toEqual(String.fromCharCode(0x80));
+    expect(cesu8ToString(highTwoByte)).toEqual(String.fromCharCode(0x7ff));
+  });
+
+  it('acts like UTF-8 for three-byte encodings', () => {
+    // 0x0800 = 0000 100000 000000 = 0x00, 0x20, 0x00
+    const lowThreeByte = Uint8Array.from([0xe0 + 0x00, 0x80 + 0x20, 0x80 + 0x00]);
+    // 0xffff = 1111 111111 111111 = 0x0f, 0x3f, 0x3f
+    const highThreeByte = Uint8Array.from([0xe0 + 0x0f, 0x80 + 0x3f, 0x80 + 0x3f]);
+    expect(cesu8ToString(lowThreeByte)).toEqual(String.fromCharCode(0x0800));
+    expect(cesu8ToString(highThreeByte)).toEqual(String.fromCharCode(0xffff));
+  });
+
+  // TODO: test the part of CESU-8 incompatible with UTF-8; however, the existing
+  //   debugger client code didn't actually support this
```

Comment:
I suspect no additional code is needed in `cesu8ToString`. The difference between CESU-8 and UTF-8 is that CESU-8 uses "surrogate pairs" to encode code points that fall outside of the basic multilingual plane (which would be encoded into 4-bytes of UTF-8 data otherwise). "Surrogate pairs" come from UTF-16, which not coincidentally ECMA/JS also uses its "language-native encoding". AFAIK, surrogate pairs is the (only) way to represent those code points outside of the basic multilingual plane in a JS string.

A simple test case would be to add convert a string with an emoji (using 😂 here as an example):

UTF-8 data:
`0xF0, 0x9F, 0x98, 0x82`

UTF-16 surrogate code points:
`0xD83D, 0xDE02`

I just tried in Node.js:
```
> String.fromCharCode(0xD83D) + String.fromCharCode(0xDE02)
'😂'
```

Reply:
I attempted to just add this emoji test, if I understood it. However, according to [wikipedia](https://en.wikipedia.org/wiki/CESU-8), there are six-byte CESU8 encodings, when the UTF-8 encoding of a surrogate point falls in the 3-byte range, and that's what I think might not be handled.

I don't fully understand it though.

Reply:
Oops my test wasn't actually working, backing it out for now.

Reply:
Added a test for it:
https://github.com/jerryscript-project/jerryscript/pull/2205/commits/44493c71eb26782e4b39c0388cf730ac1068bbcf